doc/developer/design/20210629_source_trait.md

```diff
@@ -0,0 +1,319 @@
+Since there were only three `Config` objects in the received `DataflowDesc` the
+first node is assigned the first two, and the second one the third. This
+assignment is arbitrary and can be based on more elaborate logic to balance the
+work out.
+
+The nodes are now ready to create `Instance`s from each `Config` and start
+receiving data into the dataflow. They do this using `Config::into_instance()`,
+which requires a mutable reference to a `State`. Each dataflow node constructs,
+and keeps around, a `State` object that it re-uses across all instantiations of
+a `Config`.
+
+[![https://imgur.com/8jx4oll.png](https://imgur.com/8jx4oll.png)](https://imgur.com/8jx4oll.png)
+
+```rust
+/// The trait that defines a new kind of source.
+trait Source: ToSql + FromSql {
+    /// Type holding any parameters than can be specified using SQL keywords.
+    /// For example CREATE SOURCE FROM <name> HOST 'example.com'` would
+    /// correspond to `Sql` having a field `host` of type `String`.
+    ///
+    /// Types that implement `Deserialize` automatically implement `Deserialize`
+    type Sql: for<'de> Deserialize<'de>;
+
+    /// Type used to construct instances of this source
+    type Config: InstanceConfig;
+
+    /// The name of the source
+    fn name(&self) -> &'static str;
+
+    /// Constructs a new source from the parsed SQL. This is an opportunity for the implementation
+    /// to do any sanity check of the parameters passed
+    fn from_sql(sql: Self::Sql) -> Result<Self, Error>;
+
+    /// An opportunity for the source to remove any impurities from its
+    /// definition. This will be called before permanently storing the source in
+    /// the catalog. The default implementation is a no-op
+    async fn purify(&mut self) { }
+
+    /// Constructs a new source from the parsed SQL and WITH option. This is an
+    /// opportunity for the implementation to do any sanity check of the
+    /// parameters passed
+    fn new(sql_options: Self:Options, with_options: Self::WithOptions) -> Result<Self, Error>;
+
+    /// Creates at most `workers` instance configs. These configs will be
+    /// persisted in the catalog and be used to do any associated cleanup later.
+    /// It is up to the system to decide how the returned configs will be
+    /// distributed among the dataflow workers and converted into instances
+    fn create_instances(&self, workers: usize) -> Vec<Self::Config>;
+
+    /// The schema that will be produced by this source. The `RelationDesc`
+    /// returned conveys information of the name, type, and number of columns
+    /// but also of the key, if any exists.
+    fn desc(&self) -> RelationDesc;
+
+    /// If true, the source will only be available behind the experimental flag
+    fn is_experimental(&self) -> bool {
+        false
+    }
+
+    /// If false, the source will not be available in the cloud product
+    fn is_safe(&self) -> bool {
+        true
+    }
+}
```

Comment:
Probably also want `fn is_volatile(&self) -> bool` to allow `coord` to determine whether this data can be relied upon by others to be the same.

Reply:
It may also be that the `volatility` property requires some nuance. I wouldn't be surprised if the result needed to be an `Option<Antichain<T>>` to indicate either "no guarantees" or "I have the right data from this frontier onward".